src/overlay/PeerManager.cpp

```diff
@@ -0,0 +1,503 @@
+        std::copy(std::begin(inbound), std::end(inbound),
+                  std::back_inserter(peers));
+    }
+
+    return peers;
+}
+
+std::pair<PeerRecord, bool>
+PeerManager::load(PeerBareAddress const& address)
+{
+    auto result = PeerRecord{};
+    auto inDatabase = false;
+
+    try
+    {
+        auto prep = mApp.getDatabase().getPreparedStatement(
+            "SELECT numfailures, nextattempt, type FROM peers "
+            "WHERE ip = :v1 AND port = :v2");
+        auto& st = prep.statement();
+        st.exchange(into(result.mNumFailures));
+        st.exchange(into(result.mNextAttempt));
+        st.exchange(into(result.mType));
+        std::string ip = address.getIP();
+        st.exchange(use(ip));
+        int port = address.getPort();
+        st.exchange(use(port));
+        st.define_and_bind();
+        {
+            auto timer = mApp.getDatabase().getSelectTimer("peer");
+            st.execute(true);
+            inDatabase = st.got_data();
+
+            if (!inDatabase)
+            {
+                result.mNextAttempt =
+                    VirtualClock::pointToTm(mApp.getClock().now());
+                result.mType = static_cast<int>(PeerType::INBOUND);
+            }
+        }
+    }
+    catch (soci_error& err)
+    {
+        CLOG(ERROR, "Overlay") << "PeerManager::load error: " << err.what()
+                               << " on " << address.toString();
+    }
+
+    return std::make_pair(result, inDatabase);
+}
+
+void
+PeerManager::store(PeerBareAddress const& address, PeerRecord const& peerRecord,
+                   bool inDatabase)
+{
+    std::string query;
+
+    if (inDatabase)
+    {
+        query = "UPDATE peers SET "
+                "nextattempt = :v1, "
+                "numfailures = :v2, "
+                "type = :v3 "
+                "WHERE ip = :v4 AND port = :v5";
+    }
+    else
+    {
+        query = "INSERT INTO peers "
+                "(nextattempt, numfailures, type, ip,  port) "
+                "VALUES "
+                "(:v1,         :v2,        :v3,  :v4, :v5)";
+    }
+
+    try
+    {
+        auto prep = mApp.getDatabase().getPreparedStatement(query);
+        auto& st = prep.statement();
+        st.exchange(use(peerRecord.mNextAttempt));
+        st.exchange(use(peerRecord.mNumFailures));
+        st.exchange(use(peerRecord.mType));
+        std::string ip = address.getIP();
+        st.exchange(use(ip));
+        int port = address.getPort();
+        st.exchange(use(port));
+        st.define_and_bind();
+        {
+            auto timer = mApp.getDatabase().getUpdateTimer("peer");
+            st.execute(true);
+            if (st.get_affected_rows() != 1)
+            {
+                CLOG(ERROR, "Overlay")
+                    << "PeerManager::store failed on " + address.toString();
+            }
+        }
+    }
+    catch (soci_error& err)
+    {
+        CLOG(ERROR, "Overlay") << "PeerManager::store error: " << err.what()
```

Comment:
This seems pretty lenient, as we only emit a log that database update failed. Is there a reason to not throw here? (Looks like previous version did throw in `PeerRecord::storePeerRecord`)

Reply:
throw would crash and overlay should work fine even if there are some transient database errors